packages/integrations/vue/src/editor.cts

```diff
@@ -18,29 +18,39 @@ export function toTSX(code: string, className: string): string {
 			`;
 		}
 
-		if (parsedResult.descriptor.scriptSetup) {
-			const definePropsType =
-				parsedResult.descriptor.scriptSetup.content.match(/defineProps<([\s\S]+)>/m);
+		// Vue supports 2 type of script blocks: setup and non-setup
+		const regularScriptBlockContent = parsedResult.descriptor.script?.content ?? '';
+		const { scriptSetup } = parsedResult.descriptor;
+
+		if (scriptSetup) {
+			const definePropsType = scriptSetup.content.match(/defineProps<([\S\s]+?)>\s?\(\)/m);
+			const propsGeneric = scriptSetup.attrs.generic;
+			const propsGenericType = propsGeneric ? `<${propsGeneric}>` : '';
 
 			if (definePropsType) {
 				result = `
-						${parsedResult.descriptor.scriptSetup.content}
+						${regularScriptBlockContent}
+						${scriptSetup.content}
 
-						export default function ${className}__AstroComponent_(_props: ${definePropsType[1]}): any {
+						export default function ${className}__AstroComponent_${propsGenericType}(_props: ${definePropsType[1]}): any {
 							<div></div>
 						}
 				`;
 			} else {
-				const defineProps =
-					parsedResult.descriptor.scriptSetup.content.match(/defineProps\([\s\S]+\)/m);
+				// TODO. Find a way to support generics when using defineProps without passing explicit types.
```

Comment:
Ultimately, this module shouldn't use a regex to get the types, it should use something like https://github.com/vuejs/language-tools/tree/master/packages/component-meta and https://github.com/vuejs/language-tools/tree/master/packages/component-type-helpers

which should support everything (or at least, if it doesn't, would be a bug upstream that's not in our hands)

Reply:
I agree, I have looked briefly into these packages after I opened this PR and its not so far I haven't found an easy way.

[component-meta](https://github.com/vuejs/language-tools/tree/master/packages/component-meta) gives narrows prop types and  output loses generic information, so `T extends string` becomes just string.

[component-type-helpers](https://github.com/vuejs/language-tools/tree/master/packages/component-type-helpers) needs a component instance types, to import which requires vue language server I think and still it loses all generic information and narrows types. I am not sure how would that work with virtual file.

Current approach is not sustainable and prone to fail, right now astro doesn't account for `https://vuejs.org/guide/typescript/composition-api.html#props-default-values` or experimental defineModel and so on and yet I am not sure how to get something better in a short term.